src/main/java/org/commcare/formplayer/services/MenuSessionFactory.java

```diff
@@ -60,25 +62,28 @@ public void rebuildSessionFromFrame(MenuSession menuSession) throws CommCareSess
                 EntityScreen entityScreen = (EntityScreen) screen;
                 entityScreen.init(menuSession.getSessionWrapper());
                 if (entityScreen.shouldBeSkipped()) {
-                    screen = menuSession.getNextScreen();
+                    screen = menuSession.getNextScreen(false);
                     continue;
                 }
                 SessionDatum neededDatum = entityScreen.getSession().getNeededDatum();
-                Set<String> entityIds = entityScreen.getReferenceMap().keySet();
                 for (StackFrameStep step: steps) {
                     if (step.getId().equals(neededDatum.getDataId())) {
-                        if (entityIds.contains(step.getValue())) {
-                            currentStep = step.getValue();
+                        for (TreeReference ref: entityScreen.getReferences()) {
```

Comment:
If I understand correctly, this is effectively replacing an expensive full evaluation followed by a constant time lookup with a linear lookup that in worst-case takes the same amount of time as the expensive one?

I'm inclined to say it would make sense to add that call directly to the EntityScreen itself (IE: "Confirm ID Exists as valid response" as a function call), since I think it would be valuable for that to be a clear **capability** of the screen rather than a derivative action taken against its components, unless there's a reason not to. That might let the screen itself optimize some aspects, and gets the external callers lets coupled with the internal expectations.

Reply:
Moving it to the EntityScreen makes a lot of sense to me.